idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,851 @@
+            ],
+            title=f"The following variable(s) are poorly scaled:",
+            header="=",
+            footer="=",
+        )
+
+    def display_variables_near_bounds(self, stream=stdout):
+        """
+        Prints a list of variables with values close to their bounds. Tolerance can
+        be set in the class configuration options.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        self._write_report_section(
+            stream=stream,
+            lines_list=[
+                f"{v.name}: value={value(v)} bounds={v.bounds}"
+                for v in variables_near_bounds_set(self.model)
+            ],
+            title=f"The following variable(s) have values close to their bounds:",
+            header="=",
+            footer="=",
+        )
+
+    def _check_unit_consistency(self):
+        # Check unit consistency
+        # TODO: replace once Pyomo method ready
+        # return identify_inconsistent_units(self.model)
+        inconsistent_units = ComponentSet()
+        for o in self.model.component_data_objects(
+            [Constraint, Expression, Objective], descend_into=True
+        ):
+            try:
+                assert_units_consistent(o)
+            except UnitsError:
+                inconsistent_units.add(o)
+        return inconsistent_units
+
+    def display_components_with_inconsistent_units(self, stream=stdout):
+        """
+        Prints a list of all Constraints, Expressions and Objectives in the
+        model with inconsistent units of measurement.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        self._write_report_section(
+            stream=stream,
+            lines_list=self._check_unit_consistency(),
+            title=f"The following component(s) have unit consistency issues:",
+            end_line="For more details on constraint violations, import the "
+            "assert_units_consistent method\nfrom pyomo.util.check_units",
+            header="=",
+            footer="=",
+        )
+
+    def display_constraints_with_large_residuals(self, stream=stdout):
+        """
+        Prints a list of Constraints with residuals greater than a specified tolerance.
+        Tolerance can be set in the class configuration options.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        self._write_report_section(
+            stream=stream,
+            lines_list=large_residuals_set(self.model, tol=self.residual_tolerance),
+            title=f"The following constraint(s) have large residuals:",
+            header="=",
+            footer="=",
+        )
+
+    def check_dulmage_mendelsohn_partition(self):
```

Comment:
As we are not really "checking" anything here, should this method be named something like `get_dulmage_mendelsohn_partition`?

Reply:
As an aside, is there any reason that this needs to be a method on the class, rather than a function that takes `model` as an argument? I guess the reason is that, if we wanted to optimize things later, we could have the incidence graph cached?

Reply:
Caching was one reason, but also bringing everything into a single class so it was easy to find for the user (most of the issues we got from the interviews could be summed up as "I didn't know it existed").